cpp/src/replace/replace.cu

```diff
@@ -13,31 +16,528 @@
+      auto device_out = cudf::mutable_column_device_view::create(output);
+      auto device_values_to_replace = cudf::column_device_view::create(values_to_replace);
+      auto device_replacement_values = cudf::column_device_view::create(replacement_values);
+
+      replace<<<grid.num_blocks, BLOCK_SIZE, 0, stream>>>(
+                                             *device_in,
+                                             *device_out,
+                                             valid_count,
+                                             output.size(),
+                                             *device_values_to_replace,
+                                             *device_replacement_values);
+
+      cudaStreamSynchronize(stream);
+
+      if(valid_count != nullptr){
+        cudf::size_type valids {0};
+        CUDA_TRY(cudaMemcpyAsync(&valids,
+                                 valid_count,
+                                 sizeof(cudf::size_type),
+                                 cudaMemcpyDefault,
+                                 stream));
+        output.set_null_count(output.size() - valids);
+      }
+    }
+      template <typename col_type,
+                std::enable_if_t<not cudf::is_fixed_width<col_type>()>* = nullptr>
+      void operator()(cudf::column_view const& input_col,
+                      cudf::column_view const& values_to_replace,
+                      cudf::column_view const& replacement_values,
+                      cudf::mutable_column_view& output,
+                      cudaStream_t stream = 0){
+     CUDF_FAIL("Non fixed-width types are not supported.");
+    };
+  };
+ } //end anonymous namespace
+
+namespace cudf{
+namespace detail {
+  std::unique_ptr<cudf::column> find_and_replace_all(cudf::column_view const& input_col,
+                                                     cudf::column_view const& values_to_replace,
+                                                     cudf::column_view const& replacement_values,
+                                                     rmm::mr::device_memory_resource* mr,
+                                                     cudaStream_t stream) {
+    if (0 == input_col.size() )
+    {
+      return std::make_unique<cudf::column>(input_col);
+    }
+
+    CUDF_EXPECTS(values_to_replace.size() == replacement_values.size(),
+                     "values_to_replace and replacement_values size mismatch.");
+
+    if (0 == values_to_replace.size())
+    {
+      return std::make_unique<cudf::column>(input_col);
+    }
+
+
+    CUDF_EXPECTS(input_col.type() == values_to_replace.type() &&
+                 input_col.type() == replacement_values.type(),
+                 "Columns type mismatch.");
+    CUDF_EXPECTS(values_to_replace.nullable() == false,
+                 "Nulls are in values_to_replace column.");
+
+    std::unique_ptr<column> output;
+    if (input_col.nullable() || replacement_values.nullable()) {
```

Comment:
The output only needs to be nullable if the input or replacement values actually contain null values.

Reply:
Changed.